beacon_chain/validator_client/api.nim

```
@@ -236,128 +271,192 @@ template firstSuccessParallel*(
+                      bestResponse = Opt.some(
+                        BestNodeResponse.init(node, handlerResponse, score))
+                      if perfectScore(score):
+                        perfectScoreFound = true
+                        break
+                  else:
+                    scores.add(ApiScore.init(node))
+
+              if not(isNil(timerFut)) and timerFut.finished():
+                # If timeout is exceeded we need to cancel all the tasks which
+                # are still running.
+                var pendingCancel: seq[Future[void]]
+                for future in pendingRequests.items():
+                  if not(future.finished()):
+                    pendingCancel.add(future.cancelAndWait())
+                await allFutures(pendingCancel)
+                break innerLoop
+              else:
+                if perfectScoreFound:
+                  asyncSpawn lazyWait(pendingNodes, pendingRequests, timerFut,
+                                      RequestName, strategy)
+                  break innerLoop
+                else:
+                  pendingRequests.keepItIf(it notin finishedRequests)
+                  pendingNodes.keepItIf(it notin finishedNodes)
+
+            except CancelledError as exc:
+              var pendingCancel: seq[Future[void]]
+              # `or` operation does not cancelling Futures passed as arguments.
+              if not(isNil(raceFut)) and not(raceFut.finished()):
+                pendingCancel.add(raceFut.cancelAndWait())
+              if not(isNil(timerFut)) and not(timerFut.finished()):
+                pendingCancel.add(timerFut.cancelAndWait())
+              # We should cancel all the requests which are still pending.
+              for future in pendingRequests.items():
+                if not(future.finished()):
+                  pendingCancel.add(future.cancelAndWait())
+              # Awaiting cancellations.
+              await allFutures(pendingCancel)
```

when calling cancelAndWait, is it possible that some of these futures are completed instead of cancelled?

---

Its only possible to get into this situation
So in this case:
 ```<future not finished> -> <future not finished> -> <future finished>``` result `future` will be finished, not cancelled, but it sill requires number of async steps to get into this.
And in this case
```<future not finished> -> <future not finished> -> <future cancelled>``` result `future` will be cancelled, but only if some of the intermediate futures will not decide to swallow cancellation or replace it with an error.